getnacelle/nacelle-js
# fix: use POSTs when URL too long for GETs

## Why are these changes introduced?

Addresses [ENG-9515](https://nacelle.atlassian.net/browse/ENG-9515) 

> APQ requests break when the URL exceeds 2048 characters

As described in https://github.com/urql-graphql/urql/issues/3185, [`@urql/exchange-persisted`](https://www.npmjs.com/package/@urql/exchange-persisted) currently sets urql's `preferGetMethod` setting to `'force'`. This was causing problems with an especially large query in `reference-stores/next/pages/collections/[handle].js`.

This issue is demonstrated with [`@nacelle/storefront-sdk@/2.0.0-beta.10`](https://www.npmjs.com/package/@nacelle/storefront-sdk/v/2.0.0-beta.10) in https://codesandbox.io/s/nacelle-storefront-sdk-2-0-0-demo-failure-to-fetch-with-prefergetmethod-force-2u9kjm.

## What is this pull request doing?

This PR patches [`@urql/exchange-persisted`](https://www.npmjs.com/package/@urql/exchange-persisted) using [`patch-package`](https://www.npmjs.com/package/patch-package). The patch changes the `preferGetMethod` setting from `'force'` to `'within-url-limit'` when `@urql/exchange-persisted`'s `preferGetForPersistedQueries` setting is set to `true`.

## How to Test

1. Check out the `ENG-9515/patch-urql-exchange-persisted` branch
2. `npm run bootstrap` from the monorepo root
3. Navigate to `packages/storefront-sdk`
4. `npm run build` - the package should build without issues
5. `npm run coverage` - all tests should pass with 100% coverage
6. Navigate to `reference-stores/next`
7. `npm run build` - the project should build without errors

I tested this change by publishing [`@nacelle/storefront-sdk@2.0.0-hotfix.0
`](https://www.npmjs.com/package/@nacelle/storefront-sdk/v/2.0.0-hotfix.0) with the `v2-hotfix` tag, and using it in a forked version of the bug demo Code Sandbox from above. As show in https://codesandbox.io/s/nacelle-storefront-sdk-2-0-0-demo-prefergetmethod-within-url-limits-hb0zkm?file=/src/index.ts, the request succeeds. Interestingly, it succeeds with a GET, without converting the request to a POST. I'm not yet sure why this works, but it's a good result nonetheless.

## Checklist



[ENG-9515]: https://nacelle.atlassian.net/browse/ENG-9515?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
